sycl/unittests/Extensions/WeakObject.cpp

```diff
@@ -0,0 +1,409 @@
+  sycl::accessor<int, 1, sycl::access::mode::read_write,
+                 sycl::access::target::device,
+                 sycl::access::placeholder::true_t>
+      PAcc1D1{Buf1D1};
+  sycl::accessor<int, 1, sycl::access::mode::read_write,
+                 sycl::access::target::device,
+                 sycl::access::placeholder::true_t>
+      PAcc1D2{Buf1D2};
+  sycl::accessor<int, 2, sycl::access::mode::read_write,
+                 sycl::access::target::device,
+                 sycl::access::placeholder::true_t>
+      PAcc2D1{Buf2D1};
+  sycl::accessor<int, 2, sycl::access::mode::read_write,
+                 sycl::access::target::device,
+                 sycl::access::placeholder::true_t>
+      PAcc2D2{Buf2D2};
+  sycl::accessor<int, 3, sycl::access::mode::read_write,
+                 sycl::access::target::device,
+                 sycl::access::placeholder::true_t>
+      PAcc3D1{Buf3D1};
+  sycl::accessor<int, 3, sycl::access::mode::read_write,
+                 sycl::access::target::device,
+                 sycl::access::placeholder::true_t>
+      PAcc3D2{Buf3D2};
+  sycl::accessor<int, 1, sycl::access::mode::read_write,
+                 sycl::access::target::host_buffer>
+      HAcc1D1;
+  sycl::accessor<int, 1, sycl::access::mode::read_write,
+                 sycl::access::target::host_buffer>
+      HAcc1D2;
+  sycl::accessor<int, 2, sycl::access::mode::read_write,
+                 sycl::access::target::host_buffer>
+      HAcc2D1;
+  sycl::accessor<int, 2, sycl::access::mode::read_write,
+                 sycl::access::target::host_buffer>
+      HAcc2D2;
+  sycl::accessor<int, 3, sycl::access::mode::read_write,
+                 sycl::access::target::host_buffer>
+      HAcc3D1;
+  sycl::accessor<int, 3, sycl::access::mode::read_write,
+                 sycl::access::target::host_buffer>
+      HAcc3D2;
+
+  CallableT<decltype(Ctx1)>()(Ctx1, Ctx2);
+  CallableT<decltype(Q1)>()(Q1, Q2);
+  CallableT<decltype(E1)>()(E1, E2);
+  CallableT<decltype(KB1)>()(KB1, KB2);
+  CallableT<decltype(Buf1D1)>()(Buf1D1, Buf1D2);
+  CallableT<decltype(Buf2D1)>()(Buf2D1, Buf2D2);
+  CallableT<decltype(Buf3D1)>()(Buf3D1, Buf3D2);
+  CallableT<decltype(PAcc1D1)>()(PAcc1D1, PAcc1D2);
+  CallableT<decltype(PAcc2D1)>()(PAcc2D1, PAcc2D2);
+  CallableT<decltype(PAcc3D1)>()(PAcc3D1, PAcc3D2);
+  CallableT<decltype(HAcc1D1)>()(HAcc1D1, HAcc1D2);
+  CallableT<decltype(HAcc2D1)>()(HAcc2D1, HAcc2D2);
+  CallableT<decltype(HAcc3D1)>()(HAcc3D1, HAcc3D2);
+
+  Q1.submit([&](sycl::handler &CGH) {
+    sycl::accessor<int, 1, sycl::access::mode::read> DAcc1D1{Buf1D1, CGH};
+    sycl::accessor<int, 1, sycl::access::mode::read> DAcc1D2{Buf1D2, CGH};
+    sycl::accessor<int, 2, sycl::access::mode::read> DAcc2D1{Buf2D1, CGH};
+    sycl::accessor<int, 2, sycl::access::mode::read> DAcc2D2{Buf2D2, CGH};
+    sycl::accessor<int, 3, sycl::access::mode::read> DAcc3D1{Buf3D1, CGH};
+    sycl::accessor<int, 3, sycl::access::mode::read> DAcc3D2{Buf3D2, CGH};
```

Comment:
It looks like unittests don't currently like the deduction guides, seemingly due to a problem with `IsValidTag` and `IsSameAsBuffer`. I cannot reproduce with an actual SYCL program, so I suspect it may have something to do with how the unittests are built and/or potentially a bug in the host compiler. I have a fix but I think it would be better to do in isolation as a follow-up patch.